File: app/src/main/java/bassamalim/hidaya/activities/BookViewer.java
package bassamalim.hidaya.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.SeekBar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;

import java.util.ArrayList;

import bassamalim.hidaya.R;
import bassamalim.hidaya.adapters.BookViewerAdapter;
import bassamalim.hidaya.databinding.ActivityBookViewerBinding;
import bassamalim.hidaya.models.Book;
import bassamalim.hidaya.models.BookDoorCard;
import bassamalim.hidaya.other.Utils;

public class BookViewer extends AppCompatActivity {

    private ActivityBookViewerBinding binding;
    private SharedPreferences pref;
    private int bookId;
    private int chapterId;
    private Book.BookChapter.BookDoor[] doors;
    private boolean[] favs;
    private RecyclerView recycler;
    private SeekBar textSizeSb;
    private BookViewerAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        binding = ActivityBookViewerBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        pref = PreferenceManager.getDefaultSharedPreferences(this);

        Intent intent = getIntent();
        bookId = intent.getIntExtra("book_id", 0);
        chapterId = intent.getIntExtra("chapter_id", 0);
        binding.topBarTitle.setText(intent.getStringExtra("book_title"));

        getData();

        setupRecycler();

        setupListeners();
    }

    private void getData() {
        String path = getExternalFilesDir(null) + "/Books/" + bookId  + ".json";
        String jsonStr = Utils.getJsonFromDownloads(path);
        Gson gson = new Gson();
        Book book = gson.fromJson(jsonStr, Book.class);
        doors = book.getChapters()[chapterId].getDoors();

        String favsStr = PreferenceManager.getDefaultSharedPreferences(this)
                .getString("book" + bookId + "_chapter" + chapterId + "_favs", "");
        if (favsStr.length() == 0)
            favs = new boolean[doors.length];
        else
            favs = gson.fromJson(favsStr, boolean[].class);
    }

    private ArrayList<BookDoorCard> makeCards() {
        ArrayList<BookDoorCard> cards = new ArrayList<>();
        for (int i = 0; i < doors.length; i++) {
            Book.BookChapter.BookDoor door = doors[i];
            cards.add(new BookDoorCard(door.getDoorId(), door.getDoorTitle(),
                    door.getText(), favs[i]));
        }
        return cards;
    }

    private void setupRecycler() {
        recycler = binding.recycler;
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recycler.setLayoutManager(layoutManager);
        adapter = new BookViewerAdapter(this, makeCards(), bookId, chapterId);
        recycler.setAdapter(adapter);
    }

    private void setupListeners() {
        textSizeSb = binding.textSizeSb;

        textSizeSb.setProgress(pref.getInt(getString(R.string.books_text_size_key), 15));

        binding.textSizeIb.setOnClickListener(v -> {
            if (textSizeSb.getVisibility() == View.GONE)
                textSizeSb.setVisibility(View.VISIBLE);
            else
                textSizeSb.setVisibility(View.GONE);
        });

        textSizeSb.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {}

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {}

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                SharedPreferences.Editor editor = pref.edit();
                editor.putInt(getString(R.string.books_text_size_key), seekBar.getProgress());
                editor.apply();

                adapter.setTextSize(seekBar.getProgress());
                recycler.setAdapter(null);
                recycler.setAdapter(adapter);
            }
        });
    }

    public void onDestroy() {
        super.onDestroy();
        binding = null;
        recycler.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/BookViewer.kt
package bassamalim.hidaya.activities

import android.content.SharedPreferences
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.preference.PreferenceManager
import bassamalim.hidaya.R
import bassamalim.hidaya.models.Book
import bassamalim.hidaya.ui.components.*
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.FileUtils
import com.google.gson.Gson

class BookViewer : ComponentActivity() {

    private lateinit var pref: SharedPreferences
    private val gson = Gson()
    private var bookId = 0
    private var chapterId = 0
    private lateinit var doors: Array<Book.BookChapter.BookDoor>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ActivityUtils.onActivityCreateSetLocale(this)

        pref = PreferenceManager.getDefaultSharedPreferences(this)

        val intent = intent
        bookId = intent.getIntExtra("book_id", 0)
        chapterId = intent.getIntExtra("chapter_id", 0)
        val bookTitle = intent.getStringExtra("book_title")!!

        doors = getDoors()

        setContent {
            AppTheme {
                UI(title = bookTitle)
            }
        }
    }

    private fun getDoors(): Array<Book.BookChapter.BookDoor> {
        val path = getExternalFilesDir(null).toString() + "/Books/" + bookId + ".json"
        val jsonStr = FileUtils.getJsonFromDownloads(path)
        val book = gson.fromJson(jsonStr, Book::class.java)
        return book.chapters[chapterId].doors
    }

    @Composable
    private fun UI(title: String) {
        val textSize = remember { mutableStateOf(pref.getInt(getString(R.string.alathkar_text_size_key), 15)) }

        MyScaffold(
            title = title,
            bottomBar = {
                MyReadingBottomBar(
                    textSizeState = textSize
                ) {
                    textSize.value = it.toInt()

                    pref.edit()
                        .putInt(getString(R.string.alathkar_text_size_key), it.toInt())
                        .apply()
                }
            }
        ) {
            MyLazyColumn(
                modifier = Modifier.padding(it),
                lazyList = {
                    items(items = doors) { item ->
                        DoorCard(door = item, textSize)
                    }
                }
            )
        }
    }

    @Composable
    private fun DoorCard(door: Book.BookChapter.BookDoor, textSize: MutableState<Int>) {
        val textSizeMargin = 15
        MySurface(
            Modifier.padding(vertical = 3.dp, horizontal = 5.dp),
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                MyText(
                    text = door.doorTitle,
                    modifier = Modifier.padding(10.dp),
                    fontSize = (textSize.value + textSizeMargin).sp,
                    fontWeight = FontWeight.Bold
                )

                MyText(
                    text = door.text,
                    modifier = Modifier.padding(10.dp),
                    fontSize = (textSize.value + textSizeMargin).sp,
                    textColor = AppTheme.colors.strongText
                )
            }
        }
    }

}